### PR TITLE
Improve `Tools::passwdGen` method

### PR DESCRIPTION
**Pull request by PululuK**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This change improve `Tools::passwdGen` method. I found while working on another bug..
| Type?             | refacto
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | N/A.
| How to test?      | Unit test must be happy :)
| Possible impacts? | N/A

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24027)

### CODE REVIEW

File: classes/Tools.php
Comment (PierreRambaud) [inline classes/Tools.php]:
```suggestion
```

Can be removed, it will go in default statement, otherwise the default has no aim and must raise an exception

Comment (PululuK) [reply classes/Tools.php]: Yep!

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (matks) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]: Hi @PululuK the use of `assertRegExp` can be powerful but it might not be very readable 😅  . In order to understand the test, the reader must decipher the input regexp to find out what you are trying to validate.

Could we find something that makes the test easier to read?

Readability is quite important for test because to maintain something you need to understand it. If it is hard to understand it is hard to maintain, and hard to maintain tests lose usefulness

Comment (PululuK) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]: @matks  can you please tell me what does not seem readable to you here?

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (PierreRambaud) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]: You duplicate the index `'/^(false|0)$/'`, only the last statement is executed.

I'm also afraid that you must use `false` or `0` a regex with false inside is also interpreted as string not only boolean, which mean if you have a password randomly generated the word `false` you're rekt :sweat_smile: 

Regarding the code, you should test for `false` or `0` per tests, not both together, the method can't return `false` and `0`, the if statement is not the same.

Comment (PululuK) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]: @PierreRambaud  But `false` (string) is a weak password for `openssl_random_pseudo_bytes`, so it will return `false` (bool) and not `string`.

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (matks) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]:
```
'/^(\w){10}$/' => Tools::passwdGen(10, Tools::PASSWORDGEN_FLAG_ALPHANUMERIC),
'/^(false|0)$/' => Tools::passwdGen(1),
```

These 2 statements are not obvious. When I read them first I was "gné 😮 ?"

Then I saw it was regexp, then I understood I had to convert it in my mind into something meaningful and finally I got it. However it means to read a single line took me 10 seconds.

It's not "unreadable" 😉  (this is why I added a comment but I dit not `request changes`). What I'm saying is "it could be a lot more readable"

I have some ideas to improve readability but I prefer to let you choose what is a suitable solution. [Rather than giving explicit directions I prefer to let the developer decide](https://google.github.io/eng-practices/review/reviewer/comments.html) following Google Code Review Practices 😄  .

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (matks) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]:
```suggestion
    public function passwordGenProvider()
```

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (matks) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]:
```suggestion
     * @dataProvider passwordGenProvider
```

File: classes/Tools.php
Comment (Progi1984) [inline classes/Tools.php]: Please update the PHPDoc :)

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (Progi1984) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]:
```suggestion
    public function testPasswdGen(string $expectedPassword, string $passwordGenerated): void
```

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (Progi1984) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]:
```suggestion
    public function passwodGenProvider(): array
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: classes/Tools.php
Comment (Progi1984) [reply classes/Tools.php]:
```suggestion
    public static function passwdGen($length = 8, $flag = static::PASSWORDGEN_FLAG_ALPHANUMERIC)
```

For avoiding BC Breaks

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (PululuK) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]: > I have some ideas to improve readability but I prefer to let you choose what is a suitable solution.

it is indeed very strange to hear in an exercise of code reviews.

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (Progi1984) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]: It's a data provider, so the return should be 
```suggestion
            [
                '/^(\d{8})$/',
                Tools::passwdGen(),
            ],
```
and not

```suggestion
            '/^(\d{8})$/' => Tools::passwdGen(),
```

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (matks) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]: I just want you to understand what I am going to write are _suggestions_ on how to make the code more readable, but you can _discard these suggestions_ if you find something else, than you prefer, that also improve the readability

```
self::TEN_ALPHANUMERIC => Tools::passwdGen(10, Tools::PASSWORDGEN_FLAG_ALPHANUMERIC),
self::FALSE_OR_ZERO => Tools::passwdGen(1),
```

This would be one way to make the lines easier to read 😄  . Just put a name (here a constant but it could be a variable) on the regexp. So the reader can grasp what is this line without having to decipher the regexp.

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (PululuK) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]: thanks @Progi1984, but passwordGen can return string, int or bool (false)

File: classes/Tools.php
Comment (PululuK) [reply classes/Tools.php]: default case is alphanumeric, so no BC.

Comment (Progi1984) [reply classes/Tools.php]: It's a BC as you change the signature of the method.

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (Progi1984) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]: Type hint the first parameter in this case.

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (PululuK) [reply tests-legacy/Unit/Classes/ToolsCoreTest.php]: your proposal seems very interesting, luckily you haven't kept it to yourself !
Thanks @matks

File: classes/Tools.php
Comment (Progi1984) [reply classes/Tools.php]: Ping @PululuK

File: tests-legacy/Unit/Classes/ToolsCoreTest.php
Comment (Progi1984) [inline tests-legacy/Unit/Classes/ToolsCoreTest.php]:
```suggestion
    public function testPasswdGen(string $expectedPassword, $passwordGenerated): void
```

**Review by Progi1984 — COMMENTED** [review]
Two feedbacks :p, sorry

File: classes/Tools.php
Comment (PululuK) [reply classes/Tools.php]: @Progi1984 can you give me an example of what could break something?

Comment (Progi1984) [reply classes/Tools.php]: A class which inherits of ToolsCore will be broken. Are you agree @PierreRambaud  ?

**Comment by matks** [comment]
@PululuK CI is rejecting your PR 😭 
```
1) LegacyTests\Unit\Classes\ToolsCoreTest::testPasswdGen with data set #0 (null, 'qA52hqQk')
TypeError: Argument 1 passed to LegacyTests\Unit\Classes\ToolsCoreTest::testPasswdGen() must be of the type string, null given, called in /home/runner/work/PrestaShop/PrestaShop/vendor/phpunit/phpunit/src/Framework/TestCase.php on line 1154

/home/runner/work/PrestaShop/PrestaShop/tests-legacy/Unit/Classes/ToolsCoreTest.php:378
```

**Comment by PululuK** [comment]
> @PululuK CI is rejecting your PR sob
> 
> ```
> 1) LegacyTests\Unit\Classes\ToolsCoreTest::testPasswdGen with data set #0 (null, 'qA52hqQk')
> TypeError: Argument 1 passed to LegacyTests\Unit\Classes\ToolsCoreTest::testPasswdGen() must be of the type string, null given, called in /home/runner/work/PrestaShop/PrestaShop/vendor/phpunit/phpunit/src/Framework/TestCase.php on line 1154
> 
> /home/runner/work/PrestaShop/PrestaShop/tests-legacy/Unit/Classes/ToolsCoreTest.php:378
> ```

@Progi1984  @matks  any ideia ?
The value expected is null, but if I see my data provider I have de datas  😔

File: classes/Tools.php
Comment (PrestaEdit) [inline classes/Tools.php:113]: Why not abcdefghijkmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ ?

File: classes/Tools.php
Comment (PierreRambaud) [reply classes/Tools.php]: Yes, if someone already overrides it, it will break the class.

File: classes/Tools.php
Comment (PierreRambaud) [inline classes/Tools.php]:
```suggestion
```
Not needed because of default switch.

File: classes/Tools.php
Comment (PululuK) [reply classes/Tools.php]: @PierreRambaud @Progi1984  something like http://sandbox.onlinephpfunctions.com/code/6c50a8a6e2a6ae0cb0caccc24327c35c016255d3 ?

Comment (PierreRambaud) [reply classes/Tools.php]: I forget that parameters are not typed -_-, so it doesn't break the code :sweat_smile: 
Afaik we use the Symfony Standard for Backward compatibility: https://symfony.com/doc/current/contributing/code/bc.html#id74

Regarding https://symfony.com/doc/current/contributing/code/bc.html#changing-classes change default argument value is not allowed and considered as a BC break
image

Comment (PululuK) [reply classes/Tools.php]: @PierreRambaud but I don't add a value ... I simply change the value

File: classes/Tools.php
Comment (PululuK) [reply classes/Tools.php:113]: It is outside the scope of this PR. Can you please open a new PR ?

**Review by Progi1984 — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by sarahdib** [comment]
Hello @PululuK 

Could you provide us a scenario in order to test this pr ?

Thank you :)

**Comment by PululuK** [comment]
@sarahdib I have already added different scenarios in phpunit test.
Btw @PierreRambaud  why QA is needed ?

**Comment by PierreRambaud** [comment]
> @sarahdib I have already added different scenarios in phpunit test.
> Btw @PierreRambaud why QA is needed ?

@sarahdib I'm really sorry, it's the habit to add the Waiting for QA label after an approve -_-

**Comment by PierreRambaud** [comment]
Thank you @PululuK

**Pull request merged**
